### PR TITLE
Add chisel flow and sw test for block gemm

### PR DESCRIPTION
**Pull request by xiaoling-yi**
In this PR, we add the following:

* add chisel flow using bender
* add sw test for block gemm

### CODE REVIEW

**Comment by JosseVanDelm** [comment]
@xiaoling-yi can you split this in two separate PR's please? One for adding the bender flow, and one for the software?
Also the CI is still failing :cry: let me know if you need any help!

**Comment by rgantonio** [comment]
I think we need to discuss @xiaoling-yi. I vote you add the software test first before the bender flow thingy hehe.

File: target/snitch_cluster/sw/apps/Makefile [hunk 1]
```diff
@@ -22,6 +22,7 @@ SUBDIRS += snax-mac-simple/untiled
 SUBDIRS += snax-mac-simple/tiled
 SUBDIRS += snax-gemm-engine
```
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/Makefile:23]: @xiaoling-yi  in the new PR, make sure to update this test to use the block GEMM. Basically, it's the same base GEMM test BUT using the block GEMM.

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm-load-with-stride.c
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm-load-with-stride.c:1]: CLEAN PLEZZZZ 💧 💧 💧 💧 💧 💧

Comment (rgantonio) [reply target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm-load-with-stride.c:1]: Take out the unnecessary not wanted comments.

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c:1]: Please put sufficient comments on what each function does. Maybe better if comments are in lib.h? or both? I dunno! ❓

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.h
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.h:1]: Maybe make a snax-gemm-params.h file? So that we can separate header files which are generated by params.hjson, and this snax-gemm-lib.h just contains all the functions only. Tadaaa~ might look more elegant! 🦢

File: target/snitch_cluster/sw/apps/snax-gemm-run.yaml
Comment (rgantonio) [inline target/snitch_cluster/sw/apps/snax-gemm-run.yaml:1]: Add snax-gemm-engine back in another PR but with the updated version! See issue #44

File: util/container/DockerfileChisel [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+# Copy artifacts from stage 1.
+COPY --from=builder /root/.cargo/bin/bender bin/
+COPY --from=builder /root/.cargo/bin/banshee bin/
+
+# Set locale to UTF-8, required because Python 3.6 defaults on ASCII encoding.
+# See https://click.palletsprojects.com/en/8.1.x/unicode-support/
+ENV LC_ALL "C.UTF-8"
+ENV LANG   "C.UTF-8"
+
+# Install Chisle Environment
```
Comment (rgantonio) [inline util/container/DockerfileChisel:128]: Chisel!!!

**Review by rgantonio — CHANGES_REQUESTED** [review]
In the next PR, maybe do the same thing that @JosseVanDelm did for the SNAX MAC. So have the functions and header files be declared at a higher hierarchy?

Anyway good job! I think it's not so bad 🥇

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.h [hunk 3]
```diff
@@ -0,0 +1,75 @@
+#include <stdbool.h>
+#include "snrt.h"
+#include "stdint.h"
+
+// wrap some functions for tests
+#ifndef CSR_DEF_H
```
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.h:6]: Oh, I see the `#endif` now, why do you have two separate header guards in one header file?

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                            i * meshCol + j + b * M * meshRow * meshCol * N +
+                            m * meshRow * meshCol * N + n * meshRow * meshCol;
+                        out_addr =
+                            output +
+                            (b * strideC + m * ldC + n * strideInnermostC) /
+                                sizeof(int32_t) +
+                            i * meshCol + j;
+                        // Check if output is same as golden output
+                        if ((int32_t)*out_addr != output_golden[golden_idx]) {
+                            printf("%dth not equal: output %d, golden %d \n",
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c:208]: No printfs inside your library calls. They slow down simulation!

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c
Comment (JosseVanDelm) [reply target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c:1]: You have a lot of functions here that return a bool that is hardcoded. Seems a bit pointless?
Can you just not use void types instead?

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
Hi @xiaoling-yi,

Overall this is a great PR, but it feels unfinished, unpolished and you seem to be ignoring/bypassing a lot of checks in CI! Don't do this!

File: target/snitch_cluster/Makefile [hunk 5]
```diff
@@ -82,6 +85,7 @@ VLOG_FLAGS += -suppress 13314
 VLOG_FLAGS += ${VLOG_64BIT}
 
 VLT_FLAGS  += --trace
+VLT_FLAGS  += --trace-structs
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply target/snitch_cluster/Makefile:88]: With this, we can see the hierarchy in vcd file which is convenient for debug

File: Bender.yml [hunk 6]
```diff
@@ -29,7 +29,7 @@ dependencies:
   tech_cells_generic: { git: https://github.com/pulp-platform/tech_cells_generic, version:  0.2.11  }
   riscv-dbg:          { git: https://github.com/pulp-platform/riscv-dbg,          version:  0.8.0   }
   hwpe-mac-engine:    { git: https://github.com/KULeuven-MICAS/hwpe-mac-engine.git, rev: 5d3b4525b665169fc8321c8a811f3c83ad3c72e8 }
-  snax-gemm:          { git: https://github.com/KULeuven-MICAS/snax-gemm.git,     rev: 6b538447969cb0c00a4b749b38d68ae932c6aca1 }
+  snax-gemm:          { git: https://github.com/KULeuven-MICAS/snax-gemm.git,     rev: 09fa6f1ed6d11f7fb3699fb801de02c593dade6e }
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply Bender.yml:32]: Solved!

File: target/snitch_cluster/sw/apps/snax-block-gemm/data/datagen.py
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply target/snitch_cluster/sw/apps/snax-block-gemm/data/datagen.py:1]: I ignored the line too long error. It makes more sense to put it in one line.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -50,7 +50,7 @@ jobs:
     name: Simulate SW on SNAX MAC Cluster w/ Verilator
     runs-on: ubuntu-22.04
     container:
-      image: ghcr.io/pulp-platform/snitch_cluster:ipc-verification
+      image: rgantonio/snax-chisel-test
```
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply .github/workflows/ci.yml:53]: Bender needs all files listed in Bender so it needs also the chisel-generated system verilog file. It complains if it doesn't have the sbt environment to generate it

**Comment by xiaoling-yi** [comment]
Good review! I cleaned and polished this RP a bit. 

I will add more documentation. For sw part, maybe in the next PR. For hw part, maybe in the separate GEMM repo. We can discuss how to add documentation!

File: Bender.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply Bender.yml:32]: It still seems to be pointing to
https://github.com/kuleuven-micas/snax-gemm/tree/09fa6f1ed6d11f7fb3699fb801de02c593dade6e
Which had failing unit tests 3 days ago?

File: target/snitch_cluster/Makefile (hunk 5 above)
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply target/snitch_cluster/Makefile:88]: I'm referencing the documentation here at https://veripool.org/guide/latest/exe_verilator.html#cmdoption-trace-structs.
 
> Enable tracing to show the name of packed structure, union, and packed array fields, rather than a single combined packed bus. Due to VCD file format constraints, this may result in significantly slower trace times and larger trace files.

Enabling this in CI might slow down things unnecessarily! I'd recommend to keep this somewhere if you'd like to use this personally, but we should remove this in our git repo.

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-params.h [hunk 8]
```diff
@@ -0,0 +1,8 @@
+#ifndef GEMM_PARAMETER_H
+#define GEMM_PARAMETER_H
+
+#define tileSize 8
+#define meshRow 8
+#define meshCol 8
+
+#endif
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-params.h:8]: missing a newline, you can set a setting in vscode that always inserts a newline at the bottom

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.h
Comment (JosseVanDelm) [reply target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.h:1]: You didn't address my comment!

Comment (JosseVanDelm) [reply target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.h:1]: Also please include `#pragma once` at the top here

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-params.h [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#ifndef GEMM_PARAMETER_H
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-params.h:1]: Please use `#pragma once` instead

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    int8_t* addr_b;
+    int32_t* addr_c;
+
+    bool clear;
+
+    // Read the mcycle CSR (this is our way to mark/delimit a specific code
+    // region for benchmarking)
+    uint32_t start_cycle = snrt_mcycle();
+
+    if (snrt_is_compute_core()) {
```
[the remark this answers is not in the document]
Comment (JosseVanDelm) [reply target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-lib.c:44]: Any reason this is still here?

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm.c [hunk 11]
```diff
@@ -55,37 +56,36 @@ int main() {
...
+        // Read the mcycle CSR
+        uint32_t end_cycle = snrt_mcycle();
+        printf("cycle number for CPU to do matrix multiply: %d \n",
+               end_cycle - start_cycle);
+
+        // Compare CPU result with golden model
+        err = check_result(C_cpu, C_golden, Batch, M, N, strideInnermostC, ldC,
+                           strideC);
+
+        printf("cpu err: %d\n", err);
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm.c:87]: Remove print statements please!

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm.c
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm.c:1]: There are still some print statements here.
Print statements slow down simulation and the test pipeline as a whole.
While I agree that it's important to keep track of performance, we might want to look into alternative methods of getting performance (e.g. post-processing traces instead of printing in simulation).

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
This already looks much better.
I think this PR would still benefit from trying to adhere to the suggested changes in #40 
This would keep the software side more uniform and makes it easier for me and @jorendumoulin to compile stuff for the accelerator.

I would still recommend not bypassing the linter for the python tests.
I agree with you that this is a corner case, of where the readability does not necessarily improve with shorter lines, but it's also not a big issue to just use an autoformatter like black to format it.

I'd like to not have discussions on bypassing linters too much, hence why I would still like you to remove the noqa's in the python file.

There are also some things in your PR that (might) slow down the simulation.
Please note that adding printf statements and verilator flags to increase tracing slow down the flow for all users, including those who are not debugging like you are.
I would like to keep all our tests lean and fast otherwise our infra might not scale well in the future.

File: Bender.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (xiaoling-yi) [reply Bender.yml]: Sorry, I forgot to add it! Solved now!

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-params.h (hunk 9 above)
Comment (xiaoling-yi) [reply target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-gemm-params.h]: done

**Comment by xiaoling-yi** [comment]
Everything should be fine now. But cpu can not give the right result which is strange. I tested the batch_gemm_cpu function in x86. It works fine. So I am confused now.

And the CI doesn't complain if the software test doesn't pass.
See: https://github.com/KULeuven-MICAS/snitch_cluster/actions/runs/6895539111/job/18759703568?pr=42

File: target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm.c [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+#include "snax-gemm-params.h"
+
+// gen random data
+// allocate space in TCDM
+// write data from L3 to TCDM
+// config csr
+// start
+// wait until finish
+// check result
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/src/snax-block-gemm.c:19]:
```suggestion
// This tests the following:
// 1) Generate random data with gendata.py
// 2) Allocate space in TCDM
// 3) Write data from L3 to TCDM
// 4) Configure the csrs for performing a block GEMM
// 5) Launch the accelerator
// 6) Wait until the accelerator finishes
// 7) Check the result of the CPU and the accelerator vs the golden model (gendata.py)
```

File: target/snitch_cluster/sw/snax/gemm/Makefile [hunk 13]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2023 ETH Zurich and University of Bologna.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Luca Colagrande <colluca@iis.ee.ethz.ch>
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/gemm/Makefile:5]: Change this license header to be your own

File: target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+                    addr_A = A + (b * M * meshRow * tileSize * K +
+                                  m * meshRow * tileSize * K +
+                                  k * meshRow * tileSize) /
+                                     sizeof(int8_t);
+                    snrt_dma_start_1d(addr_a, addr_A,
+                                      meshRow * tileSize * sizeof(int8_t));
+                }
+            }
+        }
+    }
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c:102]: We should change this later to 2D, 3D or even 4D data movers!

File: target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+                    // according to the strides definition
+                    addr_a = start_addr_a +
+                             (b * strideA + m * ldA + k * strideInnermostA) /
+                                 sizeof(int8_t);
+                    addr_b = start_addr_b +
+                             (b * strideB + n * ldB + k * strideInnermostB) /
+                                 sizeof(int8_t);
+                    addr_c = start_addr_c +
+                             (b * strideC + m * ldC + n * strideInnermostC) /
+                                 sizeof(int32_t);
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c:61]: What happens if this is not divisible by 4?

File: target/snitch_cluster/sw/apps/snax-block-gemm/data/params.hjson [hunk 16]
```diff
@@ -0,0 +1,30 @@
+// Copyright 2023 KU Leuven.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Xiaoling Yi <xiaoling.yi@esat.kuleuven.be>
+
+// Parameters for a block GEMM
+// need to fix batch = 1
+
+// large martix multiply 4 * 8 * 4 with no input data contention
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/data/params.hjson:10]: Are these comments still valid?

File: target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+                     uint32_t strideInnermostA, uint32_t strideInnermostB,
+                     uint32_t ldA, uint32_t ldB, uint32_t strideA,
+                     uint32_t strideB) {
+    int8_t* addr_a;
+    int8_t* addr_b;
+
+    int8_t* addr_A;
+    int8_t* addr_B;
+
+    if (snrt_is_dm_core()) {
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c:83]: Can we remove this call here, and call it from the main instead? That would greatly improve consistency!

File: target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+                                  k * meshRow * tileSize) /
+                                     sizeof(int8_t);
+                    snrt_dma_start_1d(addr_a, addr_A,
+                                      meshRow * tileSize * sizeof(int8_t));
+                }
+            }
+        }
+    }
+
+    if (snrt_is_dm_core()) {
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c:104]: Please remove this call here and call it from the main instead! Thanks!

File: target/snitch_cluster/sw/apps/snax-block-gemm/data/params.hjson [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+{
+    Batch: 1
+    M: 2,
+    N: 2,
+    K: 2,
+    strideInnermostA: 256,
+    strideInnermostB: 256,
+    strideInnermostC: 256,
```
Comment (JosseVanDelm) [inline target/snitch_cluster/sw/apps/snax-block-gemm/data/params.hjson:18]: This variable is not mentioned in your docs here:
https://github.com/KULeuven-MICAS/snax-gemm#computation-analysis-1
Neither are `delta_local_a` and `delta_local_b`?

**Review by JosseVanDelm — CHANGES_REQUESTED** [review]
Very good work!
I still have some final comments though!

File: target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c (hunk 14 above)
Comment (xiaoling-yi) [reply target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c:102]: ok!

File: target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c (hunk 15 above)
Comment (xiaoling-yi) [reply target/snitch_cluster/sw/snax/gemm/src/snax-gemm-lib.c:61]: In normal, this should be divisible by 4 because the output data type is int32. If not, the result might not be correct.

File: target/snitch_cluster/sw/apps/snax-block-gemm/data/params.hjson (hunk 19 above)
Comment (xiaoling-yi) [reply target/snitch_cluster/sw/apps/snax-block-gemm/data/params.hjson:18]: This stride is added after the doc. The new doc will include this!

**Review by rgantonio — APPROVED** [review]
I AM OKAY WITH THIS 😍 😍 😍 😍 😍 😍 😍 😍

**Review by JosseVanDelm — APPROVED** [review]
Hi @xiaoling-yi, this is a good PR, but let's try to make them a bit shorter next time!

I created #46 and #45 for some of the issues we ran into while doing this.

please go ahead and merge!
